texts/0000-empty-data.rst

```diff
@@ -0,0 +1,145 @@
+- Feature Name: Empty data types
+- Start Date: 2016-07-25
+- RFC PR:
+- Haskell Report Issue:
+
+
+
+#######
+Summary
+#######
+
+Some data types do not have any constructors, such as
+
+.. code-block:: haskell
+
+    data Void
+
+or the (in)famous
+
+.. code-block:: haskell
+
+    data State# s
+
+used in the GHC implementation of ``IO``.
+
+This proposal is about adding data types like these in a consistent, systemactic
+way.
+
+
+
+##########
+Motivation
+##########
+
+Currently, there is no way to state that a data type has no constructors at all.
+hile this might sound strange in isolation, consider a sum type like
+
+.. code-block:: haskell
+
+    data Either a b = Left a | Right b
+
+This might arise in a scenarion in which it is known that the parameter ``a`` is
+not inhabited, hence the ``Left`` case is impossible (up to ⊥). In GHC, we can
+currently (and umambiguously) define a type for this,
+
+.. code-block:: haskell
+
+    data Void
+
+without constructors (and indeed without even an ``=`` sign). Now,
+
+.. code-block:: haskell
+
+    Either Void a ~ a
```

Comment:
The use of `~` here is a bit scary. The author means "isomorphism" but `~` looks awfully like Haskell equality.